InstantTimestampTest test timezone fix

elliVM opened the pull request:
Fix #571 by setting the test cases expected timezone before running the test methods and return to system original default after

kortemik commented:
this alters the jvm state as setDefault for TimeZone will change the state globally and modifying global state in a test case is a no-go. please see https://docs.oracle.com/javase/8/docs/api/java/util/TimeZone.html#getDefault--

eemhu commented:
I wonder if it would be better to add an additional ctor to `InstantTimestamp` that allows defining the timezone. This ctor could be used in tests.

elliVM commented:
Will refactor to have a constructor with a timezone and otherwise use system default timezone

eemhu reviewed (commented):
Changes look good to me, but should the `DefaultTimeFormatTest` and `DPLTimeFormatTest` also be changed to use the new ctor?

elliVM commented:
yea I will update the other tests also

elliVM commented on src/test/java/com/teragrep/pth10/DefaultTimeFormatTest.java:
```diff
@@ -47,26 +47,23 @@
 
-    @BeforeEach
-    void enforceTimeZone() {
-        TimeZone.setDefault(TimeZone.getTimeZone("Europe/Helsinki"));
```
removed setting default TimeZone, `Europe/Helsinki` was changing depending on daylight saving time

elliVM replied:
This was actually causing the already mentioned issue of modifying the system timezone

elliVM commented on src/test/java/com/teragrep/pth10/DefaultTimeFormatTest.java:
```diff
@@ -102,16 +99,16 @@ void ISO8601_test() {
     @Test
     void defaultTimeFormatTest_2() {
         String time = "04/16/2020:10:25:40";
-        long expected = 1587021940L;
```
test date was during daylight saving time of  `GMT+3`  so removed one hour from epoch after setting all test cases to `GMT+2`

elliVM commented:
Made https://github.com/teragrep/pth_10/issues/573 for other test cases still using `TimeZone.setDefault("Europe/Helsinki")`

eemhu reviewed (approved):
builds ok, LGTM

kortemik commented on src/main/java/com/teragrep/pth10/ast/time/InstantTimestamp.java:
```diff
@@ -54,15 +54,22 @@
 
     private final String value;
     private final String timeformat;
+    private final TimeZone timeZone;
```
instant does not have a timezone by it's definition. Instant has a [atZone](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#atZone-java.time.ZoneId-)([ZoneId](https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html) zone), please consider refactoring the method that uses the timeZone to take this ZoneId/TimeZone as it's parameter.

elliVM replied:
Refactored timestamp objects to return `ZonedDateTime` from which further instant conversion can be done when required by other objects.

elliVM commented:
Encountered another issue while refactoring.
 If the value for example `2024-10-31T10:10:10Z` contains zone information like `Z` zulu time (UTC) it will take precedence over `SimpleDateFormat.setTimezone` when calling `SimpleDateFormat.parse(time)`. Meaning that the resulting instant is always for UTC and the method `ZoneId` value is ignored. This only happens if `DefaultTimeFormat` is used meaning that format is empty.
 
I will implement it to throw an logger warning in these cases but I can elevate it to an exception if we want to

elliVM commented:
Some DPL tests have time value with lower case `z` at the e.g.  `earliest=2020-12-12T00:00:00z`. I suspect it was meant to set a timezone like UTC? This was ignored by the old `SimpleDateFormat` when parsing the value. In the documentation `z` was used only in the format as a "general time zone" but in a parsed value it has no function. 
Replaced the legacy `SimpleDateFormat` with a `DateTimeFormatter`, this means that the `z` input now throws an exception instead of being ignored.

Fix?
- We can change lowercase `z` to `Z` in java and this would apply a `UTC` to these values and ignore the system default.
- Change test cases to use `Z` instead of `z`.
- Enforce the zone from the class member field (System default timezone outside of test cases). This would remove the ability to set timezones in the values when no time format is given.

elliVM commented:
rebased

elliVM commented:
Refactoring time relative objects

kortemik commented on src/main/java/com/teragrep/pth10/ast/commands/transformstatement/convert/Ctime.java:
```diff
@@ -63,15 +66,14 @@ public class Ctime implements UDF2<String, String, String> {
-
-        return formatted;
+    public String call(String epoch, String timeformat) throws Exception {
+        final ZoneId utcZoneId = ZoneId.of("UTC");
```
please create an issue about optimizing this, string conversion to time might be costly operation which could be moved to private final class member field. the same goes for the timeFormat which is unlikely to differ across the calls to the Ctime udf.

elliVM replied:
created https://github.com/teragrep/pth_10/issues/584

elliVM commented:
rebased

elliVM commented:
Refactored relative time objects

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/DefaultFormatAbsoluteTimestamp.java:
```diff
@@ -43,66 +43,68 @@
+                return absoluteTimestamp.zonedDateTime();
             }
-            catch (ParseException ignored) {
+            catch (final RuntimeException ex) {
```
why was this changed from `ParseException` to `RuntimeException`?

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/OffsetTimestamp.java:
```diff
@@ -0,0 +1,186 @@
+                throw new RuntimeException("Unsupported unit");
+        }
+
+        // ensure that year is between 1000-9999
```
should add here that it is due to an archiver limitation. it expects years to be 4 digits long

elliVM replied:
changed to IllegalArgumentException

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/RelativeOffset.java:
```diff
@@ -1,134 +0,0 @@
-            }
-        }
-        catch (ArithmeticException ae) {
-            // on overflow, check positivity/negativity and pin to max/min
```
is this case handled in the new object(s)?

elliVM replied:
added support to handle overflow situations

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/RelativeTimestamp.java:
```diff
@@ -45,41 +45,57 @@
-            time = offset.addOffset(time);
+    @Override
+    public ZonedDateTime zonedDateTime() {
+        if (isStub()) {
```
I think Stub objects are not supposed to have any functionality, so perhaps rename `isStub` to `isValid` or something else.

elliVM replied:
Correct and I think I will rename the `isStub` to `isValid` in all classes

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/SnappedTimestamp.java:
```diff
@@ -0,0 +1,286 @@
+
+    private final String timeStampString;
+    private final ZonedDateTime startTime;
+    private final Pattern subStringPattern = Pattern.compile("@([a-zA-Z])([+-]\\d+[a-zA-Z]+)");
```
the regex might not accept w0-w7 units, but it might've been like this before also

elliVM replied:
good catch I will look into accepting the week values in the regex

elliVM replied:
updated regex to support week days with digits

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/ValidOffsetAmountText.java:
```diff
@@ -0,0 +1,137 @@
+        catch (final NumberFormatException e) {
+            throw new RuntimeException("could not parse <" + input + "> to long. " + e.getMessage());
+        }
+        final long maxValue = 999999999L;
```
where does this number come from?

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/ValidRelativeTimestampText.java:
```diff
@@ -0,0 +1,81 @@
+    private final Pattern pattern;
+
+    public ValidRelativeTimestampText(final Text origin) {
+        this(origin, Pattern.compile("^(([-+])(\\d*[A-Za-z]+))?(@[A-Za-z]+([-+])?[\\dA-Za-z]*)?"));
```
same here about the w0-w7 but i would leave it for another issue since it was here before these changes

elliVM replied:
added regex support for weekdays

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/ValidTrailingRelativeTimestampText.java:
```diff
@@ -0,0 +1,91 @@
+        this.origin = origin;
+    }
+
+    public boolean isStub() {
```
same about the stub object as before

elliVM replied:
refactored to isValid

eemhu commented on src/test/java/com/teragrep/pth10/AbsoluteTimestampTest.java:
```diff
@@ -0,0 +1,160 @@
+        Assertions.assertEquals(5, zonedDateTime.getMonthValue());
+        Assertions.assertEquals(15, zonedDateTime.getDayOfMonth());
+        Assertions.assertEquals(14, zonedDateTime.getHour());
+        Assertions.assertEquals(45, zonedDateTime.getMinute());
```
might be worth also asserting the seconds, just in case.

elliVM replied:
added seconds and nano seconds

eemhu commented on src/test/java/com/teragrep/pth10/AbsoluteTimestampTest.java:
```diff
@@ -0,0 +1,160 @@
+        final String value = "2024-05-06T15:30:00+02:00";
+        final String timeformat = "yyyy-MM-dd'T'HH:mm:ssXXX";
+        final AbsoluteTimestamp timestamp = new AbsoluteTimestamp(value, timeformat, utcZone);
+        final ZonedDateTime zonedDateTime = timestamp.zonedDateTime();
```
assertions for time values?

elliVM replied:
added assertions

eemhu commented on src/test/java/com/teragrep/pth10/EarliestLatestTest.java:
```diff
@@ -472,214 +471,191 @@ public void customTimeformatTest() {
+        final long expected = startTime.plusSeconds(100).toEpochSecond();
 
-        RelativeTimeParser rtParser = new RelativeTimeParser();
         units.forEach(unit -> {
```
could add an assertion for loops executed here

elliVM replied:
assertions for loops

eemhu commented on src/test/java/com/teragrep/pth10/EarliestLatestTest.java:
```diff
@@ -472,214 +471,191 @@ public void customTimeformatTest() {
         units.forEach(unit -> {
-            String relativeTimestamp = "+" + amount + unit; //+100min etc.
-            Assertions.assertEquals(expected, rtParser.parse(relativeTimestamp).calculate(ts).getEpochSecond());
+            String relativeTimestamp = "+" + amount + unit; //+100d etc.
```
same here

eemhu commented on src/test/java/com/teragrep/pth10/EarliestLatestTest.java:
```diff
@@ -472,214 +471,191 @@ public void customTimeformatTest() {
+        final long expected = startTime.plusHours(100).toEpochSecond();
 
-        RelativeTimeParser rtParser = new RelativeTimeParser();
         units.forEach(unit -> {
```
same

eemhu commented on src/test/java/com/teragrep/pth10/OffsetTimestampTest.java:
```diff
@@ -0,0 +1,177 @@
+
+    private final ZoneId utcZone = ZoneId.of("UTC");
+    private final ZonedDateTime originTimestamp = ZonedDateTime.of(2025, 5, 15, 14, 45, 15, 790, utcZone);
+
```
I think these tests should also assert the other parts of the timestamp to make sure they don't change

elliVM replied:
added assertions for unchanged value

eemhu commented on src/test/java/com/teragrep/pth10/RelativeTimestampTest.java:
```diff
@@ -0,0 +1,142 @@
+
+    private final ZoneId utcZone = ZoneId.of("UTC");
+    private final ZonedDateTime originTimestamp = ZonedDateTime.of(2025, 5, 15, 14, 45, 15, 790, utcZone);
+
```
I think these tests should also assert the other parts of the timestamp to make sure they don't change

eemhu commented on src/test/java/com/teragrep/pth10/SnappedTimestampTest.java:
```diff
@@ -0,0 +1,397 @@
+    public void testWeekDays() {
+        final List<String> weekDays = Collections
+                .unmodifiableList(Arrays.asList("@w1", "@w2", "@w3", "@w4", "@w5", "@w6"));
+        int loops = 1;
```
I think initializing this with 0 makes more sense

elliVM replied:
changed to 0 and assertions are for `@w0` to `@w7`

eemhu commented on src/test/java/com/teragrep/pth10/SnappedTimestampTest.java:
```diff
@@ -0,0 +1,397 @@
+    public void testMonth() {
+        final SnappedTimestamp snappedTimestamp = new SnappedTimestamp("@months", originTimestamp);
+        final ZonedDateTime zonedDateTime = snappedTimestamp.zonedDateTime();
+        ;
```
unnecessary semicolon

elliVM replied:
removed

eemhu commented on src/test/java/com/teragrep/pth10/relativeTimeTest.java:
```diff
@@ -216,81 +238,6 @@ public void parseEndtimeuTest() {
-            named = "skipSparkTest",
-            matches = "true"
-    )
-    public void parseTimestampEarliestRelativeTest() {
```
these tests have been moved to some other test class?

elliVM commented on src/main/java/com/teragrep/pth10/ast/time/SnappedTimestamp.java:
```diff
@@ -0,0 +1,286 @@
+    private SnapUnit snapUnit() {
+        final String snapUnitSubstring;
+        final Matcher matcher = subStringPattern.matcher(timeStampString);
+        if (matcher.find()) {
```
created own ValidSnapToTimeText object to handle the value extraction and throws exception is not valid

elliVM commented on src/main/java/com/teragrep/pth10/ast/time/OffsetTimestamp.java:
```diff
@@ -0,0 +1,186 @@
+                updatedTime = startTime.plusYears(amount);
+                break;
+            default:
+                throw new RuntimeException("Unsupported unit");
```
changed to IllegalArgumentException

elliVM commented:
Rebased

elliVM commented:
Rebased again and set better logger levels. 
Invalid snap-to-time value after @ now throws an clear exception instead of being ignored. 
Supports week days `w0-w7` which was not the case before. `w8` and larger being invalid.
Added logging to easily trace the resulting time after all timestamps are applied. with resulting `ZonedDateTime` and epoch getting logged.

Tested most cases in QA

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/DefaultFormatAbsoluteTimestamp.java:
```diff
@@ -43,66 +43,51 @@
             }
         }
-        throw new RuntimeException("TimeQualifier conversion error: <" + time + "> can't be parsed.");
+        throw new RuntimeException("TimeQualifier conversion error: can't be parsed using default formats.");
```
this exception should include the user input for more clarity. now it might not be clear what the problem is if multiple timeQualifiers (earliest / latest) are used.

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/OffsetTimestamp.java:
```diff
@@ -0,0 +1,218 @@
+
+    private OffsetUnit offsetUnit() {
+        final String validUnitString = new ValidOffsetUnitText(offsetText).read();
+        switch (validUnitString.toLowerCase()) {
```
this might be preferred to use single point of return

eemhu commented on src/main/java/com/teragrep/pth10/ast/time/ValidOffsetUnitText.java:
```diff
@@ -0,0 +1,86 @@
+        }
+        else {
+            if (!matcher.find()) {
+                throw new RuntimeException("Text did not contain a valid offset unit");
```
perhaps also log the input

elliVM commented on src/main/java/com/teragrep/pth10/ast/time/ValidOffsetAmountText.java:
```diff
@@ -0,0 +1,134 @@
+        else if ("-".equals(amountString)) {
+            updatedString = "-1";
+        }
+        else if (amountString.matches(".*\\d.*")) {
```
It checks if there is a single digit in the string anywhere so `-1` and `1` are both true.
I could replace it with `else if (amountString.chars().anyMatch(Character::isDigit)) {`. what do you think  would that be better?

elliVM commented on src/main/java/com/teragrep/pth10/ast/time/ValidOffsetAmountText.java:
```diff
@@ -0,0 +1,134 @@
+        final long maxEpochSeconds = 999999999L;
+        final long minEpochSeconds = -999999999L;
+        if (value > maxEpochSeconds) {
+            LOGGER.info("value over max value");
```
removed these logger messages

eemhu commented:
> Added `equals()` and `hashCode` methods and EqualsVerifier tests. `ValidTrailingRelativeTimestampTextTest` could not be tested as EqualsVerifier was stuck in a recursive loop. Only solution I found was to make the class member field `Pattern` class `static` which I didn't want to do. Added `@Disable` tag for now with a message and comments about the issue. I can also add an issue about it after merge.

This was due to having the `Pattern` class imported from a wrong package.

elliVM commented:
rebased and ran spotless for updated licenses

elliVM commented:
after package rename rebase this version had some ClassDefNotFound issues in QA. will investigate

eemhu commented on src/main/java/com/teragrep/pth_10/ast/commands/evalstatement/UDFs/Mvrange.java:
```diff
@@ -102,11 +103,11 @@ else if (stepObj instanceof String) {
             // Go until incremented past end
             while (time < end) {
-                time = rtTimestamp.calculate(new Timestamp(time * 1000L)).getEpochSecond();
+                final ZonedDateTime startTime = ZonedDateTime.ofInstant(Instant.ofEpochSecond(time), ZoneId.of("UTC"));
```
I think the `ZoneId.of` call should be moved outside of the loop

elliVM replied:
moved utc `ZoneId` outside of the loop

eemhu commented on src/main/java/com/teragrep/pth_10/ast/time/RelativeTimestamp.java:
```diff
@@ -45,41 +45,78 @@
-            time = offset.addOffset(time);
+    @Override
+    public boolean isValid() {
+        boolean isStub = true;
```
change variable name to isValid

eemhu commented on src/main/java/com/teragrep/pth_10/ast/time/ValidOffsetAmountText.java:
```diff
@@ -0,0 +1,153 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(ValidOffsetAmountText.class);
+
+    private final Text origin;
+    private final Pattern pattern = Pattern.compile("^([+-]?\\d+)");
```
this could be in ctor, and main ctor could be private so the pattern cannot be changed

elliVM replied:
added constructor

eemhu commented on src/test/java/com/teragrep/pth_10/DPLTimeFormatTextTest.java:
```diff
@@ -0,0 +1,58 @@
+import org.junit.jupiter.api.Test;
+
+public class DPLTimeFormatTextTest {
+
```
should this test the read function?

elliVM replied:
added test for reads

elliVM commented on src/main/java/com/teragrep/pth_10/ast/time/ValidOffsetAmountText.java:
```diff
@@ -0,0 +1,153 @@
+                throw new RuntimeException("Matcher could not find a valid offset amount from <" + amountString + ">");
+            }
+        }
+        else {
```
simplified to remove nesting

eemhu replied:
what about the case where offsetAmount=+ (example `+h`)? Also, is it necessary that the `amountValueContainsDigits` is checked, wouldn't the regex catch that?

elliVM replied:
refactored so digits is checked only by the regex and the boolean is removed, also replaced matcher `find()` with `lookingAt()` to ensure that matching starts at the start of the string

elliVM commented on src/main/java/com/teragrep/pth_10/ast/time/OffsetTimestamp.java:
```diff
@@ -0,0 +1,249 @@
+        catch (final DateTimeException | ArithmeticException exception) {
+            LOGGER.info("Overflow exception when calculating offset <{}>", exception.getMessage());
+            if (amount > 0) {
+                LOGGER.info("Positive amount overflow, clamping to min accepted value");
```
fixed comment

eemhu reviewed (changes requested):
rebase, then should be ok

kortemik commented on src/main/java/com/teragrep/pth_10/ast/time/DPLTimestampImpl.java:
```diff
@@ -0,0 +1,136 @@
+        this(new AbsoluteTimestamp(value, timeformat, zoneId), new RelativeTimestamp(value, zoneId));
+    }
+
+    public DPLTimestampImpl(final AbsoluteTimestamp absoluteTimestamp, final RelativeTimestamp relativeTimestamp) {
```
Please describe the meaning of this ctor

elliVM replied:
this is the primary ctor to simply assign the encapsulated objects inside, however leaving it `public` could expose it to miss matching timestamp values so I refactored it to be a private ctor.

elliVM commented on src/main/java/com/teragrep/pth10/ast/commands/transformstatement/convert/Mktime.java:
```diff
@@ -60,10 +60,8 @@ public class Mktime implements UDF2<String, String, String> {
-        Long rv = format.instantOf(hrt).getEpochSecond();
-
-        return rv.toString();
+        final DPLTimestampImpl timestamp = new DPLTimestampImpl(hrt, tf);
```
refactored to full names

kortemik commented:
I think following design might make sense:

Timestamps are encapsulated into DPLTimestampString object.

The DPLTimestampString object takes DPLTimeFormats via ctor as a List.

DPLTimeFormat is an interface that provides DPLTimestamp from(String dplTimestampString)

DPLTimestampString has a method asDPLTimestamp which iterates over the provided DPLTimestamp.from() methods and collects a list of possible interpretations. If one and only one is found it returns it. If none is found it returns a stub. If multiple are found it returns IllegalStateException because multiple interpretations must not exist.

DPLTimeFormat implementation for DPLRelativeFormat takes a ctor of (String type, ZonedDateTime referencePoint) where the referencePoint is from the query start time. This must be provided within the DPLContext (for now) when the context is created. Other implementations of the DPLTimeFormat of course exist, such as the plain static formats.

elliVM commented on src/main/java/com/teragrep/pth_10/ast/DPLParserCatalystVisitor.java:
```diff
@@ -266,7 +267,8 @@ public Node visitRoot(DPLParser.RootContext ctx) {
         catCtx.setDplQuery(ctx.getText());
-
+        // set star time of query execution
+        catCtx.setStartTime(ZonedDateTime.now());
```
not 100% if this is the correct entry point to set the query start time, also should the timezone be parameterized? currently uses system default.

eemhu replied:
would also remove the need for a setter

elliVM replied:
Sounds good I'll add it to the ctor

eemhu commented on src/main/java/com/teragrep/pth_10/ast/time/DPLTimestampString.java:
```diff
@@ -0,0 +1,190 @@
+
+    private final String timestampString;
+    private final ZonedDateTime baseTime;
+    private final List<DPLTimeFormat> defaultFormats;
+    private final List<DPLTimeFormat> userDefinedFormats;
```
I think these do not need to be separate, because the parsing should fail if the user defined timeformat is not valid.

elliVM replied:
Issue was that user can provide a format where a value can match with default format and the user defined format. in that chase should we just throw an exception that they clashed?

eemhu commented on src/main/java/com/teragrep/pth_10/executor/DPLExecutorImpl.java:
```diff
@@ -65,21 +65,24 @@
     StreamingQuery streamingQuery = null;
 
-    public DPLExecutorImpl(Config config) {
+    public DPLExecutorImpl(Config config, ZonedDateTime startTime) {
```
I don't think it is possible without changes to the pth_15 component. This implementation class is loaded via `DPLExecutorFactory` that expects one `Config` type arg ctor

elliVM replied:
Then the only option currently would be to use `ZonedDateTime.now()` and system default timezone?

eemhu replied:
I think the most ideal option would be to set it in `DPLExecutorImpl.interpret()` method.

elliVM replied:
updated pth_15 version and added start time to the method

eemhu commented on src/main/java/com/teragrep/pth_10/ast/time/OffsetTimestamp.java:
```diff
@@ -0,0 +1,254 @@
+        return false;
+    }
+
+    private long offsetAmount() {
+        final String validAmountString = new ValidOffsetAmountText(offsetText).read();
+        return Long.parseLong(validAmountString);
+    }
+
+    private OffsetUnit offsetUnit() {
+        final String validUnitString = new ValidOffsetUnitText(offsetText).read();
```
maybe these two should be their own objects?

eemhu commented on src/test/java/com/teragrep/pth_10/DPLTimestampStringTest.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Test
+    public void testUserDefinedFormat() {
```
test invalid user defined format also

elliVM commented:
Rebased
